File: src/main/graaljs/net/x3djsonld/data/flower3.js
load('X3Dautoclass.js');

// Javadoc annotations follow, see below for source.
/**
 * <p> a flower. </p>
 <p> Related links: flower3.java source, <a href="https://www.web3d.org/x3d/content/examples/X3dResources.html" target="_blank">X3D Resources</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dSceneAuthoringHints.html" target="_blank">X3D Scene Authoring Hints</a> and <a href="https://www.web3d.org/x3d/content/X3dTooltips.html" target="_blank">X3D Tooltips</a>. </p>
	<table style="color:black; border:0px solid; border-spacing:10px 0px;" summary="Scene Metadata">
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center; padding:10px 0px;"><i>meta tags</i></td>
			<td style="text-align:left;   padding:10px 0px;">net.x3djsonld.data.flower3&nbsp; Document Metadata </td>
		</tr>

		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> title </i> </td>
			<td> <a href="https://coderextreme.net/X3DJSONLD/src/main/data/flower3.x3d">flower3.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> creator </i> </td>
			<td> John Carlson </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> generator </i> </td>
			<td> manual </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> identifier </i> </td>
			<td> <a href="https://coderextreme.net/X3DJSONLD/src/main/data/flower3.x3d" target="_blank">https://coderextreme.net/X3DJSONLD/src/main/data/flower3.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> description </i> </td>
			<td> a flower </td>
		</tr>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center;" colspan="2">  &nbsp; </td>
		</tr>
	</table>

	<p>
		This program uses the
		<a href="https://www.web3d.org/specifications/java/X3DJSAIL.html" target="_blank">X3D Java Scene Access Interface Library (X3DJSAIL)</a>.
		It has been produced using the 
		<a href="https://www.web3d.org/x3d/stylesheets/X3dToES5.xslt" target="_blank">X3dToES5.xslt</a>
		stylesheet to create ES5 source code from an <code>.x3d</code> scene.
	</p>

	* @author John Carlson
 */

function flower3
  /** Default constructor to create this object. */
  ()
  {
    /** The initialized model object, created within initialize() method. */
    this.x3dModel = {};
    this.initialize();
    return this;
  }
  function doubleToFloat(d) {
    if (Float32Array) {
        return new Float32Array([d])[0];
    }
  }
flower3.prototype = {
  /** Create and initialize the X3D model. */
  initialize : function ()
  {
  this.x3dModel = new X3D().setProfile("Immersive").setVersion("4.0")
  .setHead(new head()
    .addMeta(new meta().setName("title").setContent("flower3.x3d"))
    .addMeta(new meta().setName("creator").setContent("John Carlson"))
    .addMeta(new meta().setName("generator").setContent("manual"))
    .addMeta(new meta().setName("identifier").setContent("https://coderextreme.net/X3DJSONLD/src/main/data/flower3.x3d"))
    .addMeta(new meta().setName("description").setContent("a flower")))
  .setScene(new Scene()
    .addChild(new NavigationInfo())
    .addChild(new DirectionalLight().setDirection(doubleToFloat(0.0),doubleToFloat(-0.8),doubleToFloat(-0.2)).setIntensity(0.5))
    .addChild(new Background().setSkyColor(new MFColor(Java.to([doubleToFloat(1.000),doubleToFloat(1.000),doubleToFloat(1.000)], Java.type("float[]")))))
    .addChild(new Viewpoint().setDescription("One mathematical orbital").setPosition(doubleToFloat(0.0),doubleToFloat(0.0),doubleToFloat(50.0)))
    .addChild(new Transform("OrbitTransform").setTranslation(doubleToFloat(8.0),doubleToFloat(0.0),doubleToFloat(0.0))
      .addChild(new Shape()
        .setAppearance(new Appearance()
          .setMaterial(new Material().setDiffuseColor(doubleToFloat(0.0),doubleToFloat(0.5),doubleToFloat(1.0)).setSpecularColor(doubleToFloat(0.0),doubleToFloat(0.5),doubleToFloat(1.0))))
        .setGeometry(new IndexedFaceSet("Orbit").setDEF("Orbit").setConvex(false)
          .setCoord(new Coordinate("OrbitCoordinates")))))
    .addChild(new Script("OrbitScript").setSourceCode("\n" + 
"ecmascript:" + "\n" + 
"    " + "\n" + 
"var e = 5;" + "\n" + 
"var f = 5;" + "\n" + 
"var g = 5;" + "\n" + 
"var h = 5;" + "\n" + 
"var resolution = 100;" + "\n" + 
"var t = 0;" + "\n" + 
"var p = 0;" + "\n" + 
"\n" + 
"function initialize() {" + "\n" + 
"     generateCoordinates(resolution);" + "\n" + 
"     var localci = [];" + "\n" + 
"     for ( i = 0; i < resolution-1; i++) {" + "\n" + 
"     	for ( j = 0; j < resolution-1; j++) {" + "\n" + 
"	     localci.push(i*resolution+j);" + "\n" + 
"	     localci.push(i*resolution+j+1);" + "\n" + 
"	     localci.push((i+1)*resolution+j+1);" + "\n" + 
"	     localci.push((i+1)*resolution+j);" + "\n" + 
"	     localci.push(-1);" + "\n" + 
"	}" + "\n" + 
"    }" + "\n" + 
"    coordIndexes = new MFInt32(localci);" + "\n" + 
"}" + "\n" + 
"\n" + 
"function generateCoordinates(resolution) {" + "\n" + 
"     var theta = 0.0;" + "\n" + 
"     var phi = 0.0;" + "\n" + 
"     var delta = (2 * 3.141592653) / (resolution-1);" + "\n" + 
"     var localc = [];" + "\n" + 
"     for (var  i = 0; i < resolution; i++) {" + "\n" + 
"     	for (var  j = 0; j < resolution; j++) {" + "\n" + 
"		var rho = e + f * Math.cos(g * theta) * Math.cos(h * phi);" + "\n" + 
"		localc.push(new SFVec3f(" + "\n" + 
"			rho * Math.cos(phi) * Math.cos(theta)," + "\n" + 
"			rho * Math.cos(phi) * Math.sin(theta)," + "\n" + 
"			rho * Math.sin(phi)" + "\n" + 
"		));" + "\n" + 
"		theta += delta;" + "\n" + 
"	}" + "\n" + 
"	phi += delta;" + "\n" + 
"     }" + "\n" + 
"     coordinates = new MFVec3f(localc);" + "\n" + 
"}" + "\n" + 
"\n" + 
"function set_fraction(fraction, eventTime) {" + "\n" + 
"	t += 0.5;" + "\n" + 
"	p += 0.5;" + "\n" + 
"	var choice = Math.floor(Math.random() * 4);" + "\n" + 
"	switch (choice) {" + "\n" + 
"	case 0:" + "\n" + 
"		e += Math.floor(Math.random() * 2) * 2 - 1;" + "\n" + 
"		break;" + "\n" + 
"	case 1:" + "\n" + 
"		f += Math.floor(Math.random() * 2) * 2 - 1;" + "\n" + 
"		break;" + "\n" + 
"	case 2:" + "\n" + 
"		g += Math.floor(Math.random() * 2) * 2 - 1;" + "\n" + 
"		break;" + "\n" + 
"	case 3:" + "\n" + 
"		h += Math.floor(Math.random() * 2) * 2 - 1;" + "\n" + 
"		break;" + "\n" + 
"	}" + "\n" + 
"	if (f < 1) {" + "\n" + 
"		f = 10;" + "\n" + 
"	}" + "\n" + 
"	if (g < 1) {" + "\n" + 
"		g = 4;" + "\n" + 
"	}" + "\n" + 
"	if (h < 1) {" + "\n" + 
"		h = 4;" + "\n" + 
"	}" + "\n" + 
"	generateCoordinates(resolution);" + "\n" + 
"}" + "\n")
      .addField(new field().setName("set_fraction").setType("SFFloat").setAccessType("inputOnly"))
      .addField(new field().setName("coordinates").setType("MFVec3f").setAccessType("outputOnly"))
      .addField(new field().setName("coordIndexes").setType("MFInt32").setAccessType("outputOnly"))
      .addComments(Java.to(["",
"        <field accessType=\"inputOutput\" name=\"e\" type=\"SFFloat\" value=\"5\"/>",
"        <field accessType=\"inputOutput\" name=\"f\" type=\"SFFloat\" value=\"5\"/>",
"        <field accessType=\"inputOutput\" name=\"g\" type=\"SFFloat\" value=\"5\"/>",
"        <field accessType=\"inputOutput\" name=\"h\" type=\"SFFloat\" value=\"5\"/>",
"        <field accessType=\"inputOutput\" name=\"t\" type=\"SFFloat\" value=\"0\"/>",
"        <field accessType=\"inputOutput\" name=\"p\" type=\"SFFloat\" value=\"0\"/>",
"        <field accessType=\"initializeOnly\" name=\"resolution\" type=\"SFInt32\" value=\"100\"/>"], Java.type("java.lang.String[]"))))
    .addChild(new TimeSensor("Clock").setCycleInterval(16).setLoop(true))
    .addChild(new ROUTE().setFromNode("OrbitScript").setFromField("coordIndexes").setToNode("Orbit").setToField("set_coordIndex"))
    .addChild(new ROUTE().setFromNode("OrbitScript").setFromField("coordinates").setToNode("OrbitCoordinates").setToField("point"))
    .addChild(new ROUTE().setFromNode("Clock").setFromField("fraction_changed").setToNode("OrbitScript").setToField("set_fraction")));
  },
  // end of initialize() method


  /** Provide a shallow copy of the X3D model.
   * @return flower3 model
   */
  getX3dModel : function()
  {	  
	  return this.x3dModel;
  },
  
  /** Indicate X3DJSAIL validation results for this X3D model.
   * @return validation results plus exception information, if any
   */
  validateSelf : function()
  {
	var       metaResult = "";
	var validationResult = "";
	var  exceptionResult = "";
	try
	{
		this.initialize();
		
		if ((this.getX3dModel() == null) || (this.getX3dModel().getHead() == null))
		{
			validationResult = "empty scene, nothing to validate. " + this.x3dModel.validate();
			return validationResult;
		}
		// first list informational meta elements of interest
		var metaList = this.getX3dModel().getHead().getMetaList();
		for (var m in metaList) {
			meta = metaList[m];
			if (meta.getName().equals(meta.NAME_ERROR) ||
				meta.getName().equals(meta.NAME_WARNING) ||
				meta.getName().equals(meta.NAME_HINT) ||
				meta.getName().equals(meta.NAME_INFO) ||
				meta.getName().equals(meta.NAME_TODO))
			{
				metaResult += meta.toStringX3D();
			}
		}
		validationResult += this.x3dModel.validate(); // walk entire tree to validate correctness
	}
	catch (e)
	{
		exceptionResult = e; // report exception failures, if any
	}
	if  (metaResult === "" && exceptionResult === "" && validationResult === "")
	     return "success";
	else
	{
		var returnMessage = metaResult;
		if  (exceptionResult !== "" && validationResult !== "")
			returnMessage += "\n*** ";
		returnMessage += exceptionResult;
		if  (exceptionResult === "" && validationResult !== "")
			returnMessage = "\n" + returnMessage; // skip line before meta tags, etc.
		returnMessage += validationResult;
		return returnMessage;
	}
  },
    /** Default main() method provided for test purposes.
     * @param argv input parameters
     */
    main : function (argv)
    {
		var testObject = new flower3();
		print ("flower3 execution self-validation test results: " + testObject.validateSelf());
	}
}
new flower3().main();

File: src/main/graalpy/net/coderextreme/data/flower3.py
import x3dpsail as x3d

x3d.ConfigurationProperties.setXsltEngine(x3d.ConfigurationProperties.XSLT_ENGINE_NATIVE_JAVA)
x3d.ConfigurationProperties.setDeleteIntermediateFiles(False)
x3d.ConfigurationProperties.setStripTrailingZeroes(True)
x3d.ConfigurationProperties.setStripDefaultAttributes(True)
X3D0 = x3d.X3D()
X3D0.setProfile("Immersive")
X3D0.setVersion("4.0")
head1 = x3d.head()
meta2 = x3d.meta()
meta2.setName("title")
meta2.setContent("flower3.x3d")

head1.addMeta(meta2)
meta3 = x3d.meta()
meta3.setName("creator")
meta3.setContent("John Carlson")

head1.addMeta(meta3)
meta4 = x3d.meta()
meta4.setName("generator")
meta4.setContent("manual")

head1.addMeta(meta4)
meta5 = x3d.meta()
meta5.setName("identifier")
meta5.setContent("https://coderextreme.net/X3DJSONLD/src/main/data/flower3.x3d")

head1.addMeta(meta5)
meta6 = x3d.meta()
meta6.setName("description")
meta6.setContent("a flower")

head1.addMeta(meta6)

X3D0.setHead(head1)
Scene7 = x3d.Scene()
NavigationInfo8 = x3d.NavigationInfo()

Scene7.addChild(NavigationInfo8)
DirectionalLight9 = x3d.DirectionalLight()
DirectionalLight9.setDirection(x3d.doubleToFloat([0,-0.8,-0.2]))
DirectionalLight9.setIntensity(0.5)

Scene7.addChild(DirectionalLight9)
Background10 = x3d.Background()
Background10.setSkyColor(x3d.doubleToFloat([1,1,1]))

Scene7.addChild(Background10)
Viewpoint11 = x3d.Viewpoint()
Viewpoint11.setDescription("One mathematical orbital")
Viewpoint11.setPosition(x3d.doubleToFloat([0,0,50]))

Scene7.addChild(Viewpoint11)
Transform12 = x3d.Transform()
Transform12.setDEF("OrbitTransform")
Transform12.setTranslation(x3d.doubleToFloat([8,0,0]))
Shape13 = x3d.Shape()
Appearance14 = x3d.Appearance()
Material15 = x3d.Material()
Material15.setDiffuseColor(x3d.doubleToFloat([0,0.5,1]))
Material15.setSpecularColor(x3d.doubleToFloat([0,0.5,1]))

Appearance14.setMaterial(Material15)

Shape13.setAppearance(Appearance14)
IndexedFaceSet16 = x3d.IndexedFaceSet()
IndexedFaceSet16.setConvex(False)
IndexedFaceSet16.setDEF("Orbit")
Coordinate17 = x3d.Coordinate()
Coordinate17.setDEF("OrbitCoordinates")

IndexedFaceSet16.setCoord(Coordinate17)

Shape13.setGeometry(IndexedFaceSet16)

Transform12.addChild(Shape13)

Scene7.addChild(Transform12)
Script18 = x3d.Script()
Script18.setDEF("OrbitScript")
field19 = x3d.field()
field19.setName("set_fraction")
field19.setAccessType("inputOnly")
field19.setType("SFFloat")

Script18.addField(field19)
field20 = x3d.field()
field20.setName("coordinates")
field20.setAccessType("outputOnly")
field20.setType("MFVec3f")

Script18.addField(field20)
field21 = x3d.field()
field21.setName("coordIndexes")
field21.setAccessType("outputOnly")
field21.setType("MFInt32")

Script18.addField(field21)

Script18.addComments(x3d.CommentsBlock("""<field accessType=\"inputOutput\" name=\"e\" type=\"SFFloat\" value=\"5\"/>"""))

Script18.addComments(x3d.CommentsBlock("""<field accessType=\"inputOutput\" name=\"f\" type=\"SFFloat\" value=\"5\"/>"""))

Script18.addComments(x3d.CommentsBlock("""<field accessType=\"inputOutput\" name=\"g\" type=\"SFFloat\" value=\"5\"/>"""))

Script18.addComments(x3d.CommentsBlock("""<field accessType=\"inputOutput\" name=\"h\" type=\"SFFloat\" value=\"5\"/>"""))

Script18.addComments(x3d.CommentsBlock("""<field accessType=\"inputOutput\" name=\"t\" type=\"SFFloat\" value=\"0\"/>"""))

Script18.addComments(x3d.CommentsBlock("""<field accessType=\"inputOutput\" name=\"p\" type=\"SFFloat\" value=\"0\"/>"""))

Script18.addComments(x3d.CommentsBlock("""<field accessType=\"initializeOnly\" name=\"resolution\" type=\"SFInt32\" value=\"100\"/>"""))

Script18.setSourceCode('''ecmascript:

var e = 5;
var f = 5;
var g = 5;
var h = 5;
var resolution = 100;
var t = 0;
var p = 0;

function initialize() {
     generateCoordinates(resolution);
     var localci = new MFInt32();
     var arrIndex = 0;
     for (var i = 0; i < resolution-1; i++) {
     	for (var j = 0; j < resolution-1; j++) {
	     localci[arrIndex++] = i*resolution+j;
	     localci[arrIndex++] = i*resolution+j+1;
	     localci[arrIndex++] = (i+1)*resolution+j+1;
	     localci[arrIndex++] = (i+1)*resolution+j;
	     localci[arrIndex++] = -1;
	}
    }
    coordIndexes = localci;
}

function generateCoordinates(resolution) {
     var theta = 0.0;
     var phi = 0.0;
     var delta = (2 * 3.141592653) / (resolution-1);
     var localc = new MFVec3f();
     var arrIndex = 0;
     for (var  i = 0; i < resolution; i++) {
     	for (var  j = 0; j < resolution; j++) {
		var rho = e + f * Math.cos(g * theta) * Math.cos(h * phi);
		localc[arrIndex++] = new SFVec3f(
			rho * Math.cos(phi) * Math.cos(theta),
			rho * Math.cos(phi) * Math.sin(theta),
			rho * Math.sin(phi)
		);
		theta += delta;
	}
	phi += delta;
     }
     coordinates = localc;
}

function set_fraction(fraction, eventTime) {
	t += 0.5;
	p += 0.5;
	var choice = Math.floor(Math.random() * 4);
	switch (choice) {
	case 0:
		e += Math.floor(Math.random() * 2) * 2 - 1;
		break;
	case 1:
		f += Math.floor(Math.random() * 2) * 2 - 1;
		break;
	case 2:
		g += Math.floor(Math.random() * 2) * 2 - 1;
		break;
	case 3:
		h += Math.floor(Math.random() * 2) * 2 - 1;
		break;
	}
	if (f < 1) {
		f = 10;
	}
	if (g < 1) {
		g = 4;
	}
	if (h < 1) {
		h = 4;
	}
	generateCoordinates(resolution);
}''')

Scene7.addChild(Script18)
TimeSensor22 = x3d.TimeSensor()
TimeSensor22.setDEF("Clock")
TimeSensor22.setCycleInterval(16)
TimeSensor22.setLoop(True)

Scene7.addChild(TimeSensor22)
ROUTE23 = x3d.ROUTE()
ROUTE23.setFromNode("OrbitScript")
ROUTE23.setFromField("coordIndexes")
ROUTE23.setToNode("Orbit")
ROUTE23.setToField("set_coordIndex")

Scene7.addChild(ROUTE23)
ROUTE24 = x3d.ROUTE()
ROUTE24.setFromNode("OrbitScript")
ROUTE24.setFromField("coordinates")
ROUTE24.setToNode("OrbitCoordinates")
ROUTE24.setToField("point")

Scene7.addChild(ROUTE24)
ROUTE25 = x3d.ROUTE()
ROUTE25.setFromNode("Clock")
ROUTE25.setFromField("fraction_changed")
ROUTE25.setToNode("OrbitScript")
ROUTE25.setToField("set_fraction")

Scene7.addChild(ROUTE25)

X3D0.setScene(Scene7)
X3D0.toFileX3D("../data/flower3.new.graalpy.x3d")
X3D0.toFileJSON("../data/flower3.new.graalpy.x3dj")
